### PR TITLE
Initial pass of adding support for color displays

### LINKED ISSUE #15: Support color displays
As a first step, I imagine we should figure out the styling granularity we want to support, and make it work with BinaryColor.

One implementation detail that might pose a difficulty is that the way we invert pixel values for the single-touch interaction's input progress might not be easy to get working with arbitrary colors. We might want to do something else instead, like modify the overlay to use two distinct color values instead of inverting.

I  personally don't use color displays so I'm not the most motivated person to work on this feature.

### PR DESCRIPTION
**Pull request by BroderickCarlin**
This PR is meant as a first-pass at adding support for color displays, including a new `color.rs` example illustrating basic color support for the menu items and indicator. All other examples should continue to work as-is and existing code should continue to work with minimal to no modifications. 

Currently, the macros have not been updated and are broken within this branch. Updating these felt like a significant lift and it was noticed that there is a mention of removing macros withing the [0.6 planning issue](https://github.com/bugadani/embedded-menu/issues/36). There is also a bit of commented out code related to the inversion selector - given that "inversion" is not a universally support operation within all color palettes, and there is no existing trait (afaik) to support this operation. 

This is very light in terms of features for color support, but aims to be more of a stepping stone to adding richer color menus. At this point I am mostly curious to collect feedback

note: this PR currently includes a patch to a local copy of `embedded-text` utilizing a recent and unreleased change to avoid propagation of `From<Rgb888>` trait bounds - this obviously would be removed before any potential merge

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -118,39 +118,42 @@ where
...
 
-impl<C, R> MenuStyle<C, LineIndicator, Programmed, StaticPosition, R>
+impl<IC, MC, R> MenuStyle<LineIndicator<IC>, Programmed, StaticPosition, R, MC>
 where
-    C: PixelColor,
+    IC: PixelColor,
+    MC: PixelColor,
 {
-    pub const fn new(color: C) -> Self {
+    pub const fn new(menu_color: MC, indicator_color: IC) -> Self {
```
Comment (bugadani) [inline src/lib.rs]: I'm not sure why we need two color types. I don't think different types would compile for IC and MC?

So what I think we should do is define a Theme of sorts, instead of inserting every color one-by-one here.

```rust
trait Theme {
    type Color: PixelColor;

    fn text_color(&self) -> Self::Color;
    fn inverted_text_color(&self) -> Self::Color; // Inverted referring to the text under the progress display part of the indicator
    fn indicator_color() -> Self::Color;
... we add more as needed
}
```

This would be easy to implement for BinaryColor by default and maybe we could provide named Themes for other color types.

File: src/selection_indicator/mod.rs [hunk 2]
```diff
@@ -200,15 +199,16 @@ where
...
             &menu_state.indicator_state.state,
             input_state,
-            &mut display.cropped(&selected_item_area),
+            &mut display.cropped(&selected_item_area).color_converted(),
         )?;
 
-        // Translate inverting area to its position
-        let invert_area = invert_area.translate(selected_item_area.top_left);
-        let mut inverting = display.invert_area(&invert_area);
+        // // TODO: Figure out how to handle inversion
```
Comment (bugadani) [inline src/selection_indicator/mod.rs:208]: I think we could move this problem into the Theme. Some themes may not need any color manipulation and it may make readability worse. They could define a noop conversion method.

Inverting is just making the text visible under a background that is filling in with the same color on a BinaryColor display. I think we could:

 - keep text rendering BinaryColor and treat the text it draws as a mask
 - modify the adapter to take two colors
 - map these two colors to the text's On and Off colors, or flip the mapping over the inverting area

**Review by bugadani — COMMENTED** [review]
Thanks for working on this! I have a few ideas, I'd like to know what you think about them.

First, and most important is that we'll need to add a fancy picture to the readme :)

File: src/lib.rs (hunk 1 above)
Comment (BroderickCarlin) [reply src/lib.rs]: This is admittedly an artifact of an original pass attempting to be as flexible with color spaces as possible on a per-style component basis. These two colors can be different, with the only restriction being that each can be converted into the color space used by the display. 

A lot of this work was walked back as it felt unnecessarily complex and was a larger change than I think was actually necessary or desired, especially for this first pass. I'd be happy to switch these both to use the same color space

File: src/selection_indicator/mod.rs (hunk 2 above)
Comment (BroderickCarlin) [reply src/selection_indicator/mod.rs:208]: > Inverting is just making the text visible under a background that is filling in with the same color on a BinaryColor display.

I wonder if it is really more simple than this - and simply swapping the color of the background with the color of the text. The mask approach feels unnecessarily complex, unless there is a desire to support menus with transparent backgrounds. Personally, that feels a bit more like a niche use case but I also admit I don't currently have any grasp on how this project may be currently used by others in this regard.

**Comment by BroderickCarlin** [comment]
With the addition of color support, I think one topic that may need to be discussed is how styles are handled. It seems that color support opens the door for a much deeper and richer customization of the menu display style. I'm not sure at what scope this crate hopes to support customization, but getting a better understanding of what that aspect looks like may help some of the other pieces more naturally fall into place (such as, the topic of selection color inversion). Curious for your thoughts here, as I believe this can help to put up some clear guardrails on these conversations and this addition as a whole 😄

File: src/lib.rs (hunk 1 above)
Comment (bugadani) [reply src/lib.rs]: I have to admit I don't see a point to two color spaces for two different elements on the screen. It will all boil down to a single color in the end and conversion steps in-between would just make things more complicated than necessary. I hardly see reason to allow setting text color to `BinaryColor::On` with an RGB888 indicator color, even though some conversion could probably make something of these two. From a user's perspective this is very confusing.

File: src/selection_indicator/mod.rs (hunk 2 above)
Comment (bugadani) [reply src/selection_indicator/mod.rs:208]: I'm not sure what you're arguing for here. This "inversion" that is now commented out is BinaryColor-specific: if you have a text color, and a background color, then if the background color changes under the text, the text color has to be inverted for the text to be readable.

With a color display, inversion is:
 - not always the best option
 - not always necessary as you can have two different background colors and a third one for the text that looks always good

An adapter that would map the text pixels vs empty pixels to one color on the other would support both use cases. Text is already drawn as a bitmask mapped to different colors by embedded-graphics, just without any "background-awareness" that we need here, meaning that sticking to BinaryColor as our text renderer's direct output would save an intermediate conversion.

**Comment by bugadani** [comment]
> I'm not sure at what scope this crate hopes to support customization

It's up for debate, I'm using monochrome displays at the moment where the only real possibility is to pick what the foreground color is, the rest falls out of that automatically for me.

I'm trying to push the Theme approach as it goes relatively well to both BinaryColor and RGB, as well as it should give us an easy want to split out a color for some UI element if we need to. It's just my first idea and I often stick to those more than I should, so I'd love to hear alternatives.

File: src/lib.rs (hunk 1 above)
Comment (BroderickCarlin) [reply src/lib.rs]: very fair point. Where my mind was originally at with this was looking to allow people to write crates that act as extensions of `embedded-menu` through implementation of the `MenuItem` trait. However, I feel this comes with multiple undesirable side effects such as complicating the code within `embedded-menu` as well as allowing extension crate authors to be lazy with their menu item's color space. 

In short - I think this PR can be updated to resolve to a single mono-color space

File: src/selection_indicator/mod.rs (hunk 2 above)
Comment (BroderickCarlin) [reply src/selection_indicator/mod.rs:208]: What I was suggesting is that if a menu's color is stylized through something akin to a "theme" (as mentioned elsewhere in this PR), and the "theme" specifies both a text color and a background color, an ignorant process of "inversion" becomes just swapping these two when drawing the specified `MenuItem`. 

I suspect if the "theme" approach is taken that it may just make more sense to specify a "selected color", or such. 

EDIT: I see now that your proposed `Theme` trait even has a method to get the inverted theme color - I suspect we really are saying the same thing

**Comment by BroderickCarlin** [comment]
The `Theme` trait that you've outlined [here](https://github.com/bugadani/embedded-menu/pull/38#discussion_r1375722450) aligns pretty closely with my own thoughts.

Just looking over things quickly and it looks like a non-exhaustive list of the changes would be:
- `MenuStyle` would need to be updated to specify a `Theme` rather than a `Color`
- `Indicator` would lose its internal style, and have relevant values passed in by the `MenuStyle`

I feel I must be missing something, but it appears most things already are wired to work pretty well as-is if the `MenuStyle` change is added. 

Curious for your thoughts here and how far you'd like to see this PR go or if you'd like to see things more broken up

**Comment by bugadani** [comment]
> Curious for your thoughts here and how far you'd like to see this PR go or if you'd like to see things more broken up

I think I would be happy to merge this if you combine the color parameters into one and after you rebase. Anything beyond that is up to you.

**Comment by bugadani** [comment]
First up, I apologize, I only meant merging the color types, not necessarily the color values. Anyway, we can build on this later, so thanks for updating the PR!

One minor thing, you left your local patch in the Cargo.toml :)

**Comment by BroderickCarlin** [comment]
Made the change to reduce down to a single color parameter. At this point it is just reliant upon a release of `embedded-text` to pull in the change to remove `RGB888` conversion bounds

**Comment by bugadani** [comment]
You're right, sorry. Published as `embedded-text 0.7.0` just now

**Comment by bugadani** [comment]
Thank you very much!
